src/readbin/byteweight_util.ml

```diff
@@ -64,42 +64,54 @@ let train meth length comp db paths =
   printf "Signatures are stored in %s\n%!" db;
   Ok ()
 
-let find threshold length comp path input : unit t =
+let output_sexp img bw length threshold =
+  let fs_set = Table.foldi (Image.sections img) ~init:Addr.Set.empty
+      ~f:(fun mem sec fs_s ->
+          if Section.is_executable sec then
+            let new_fs_s = BW.find bw ~length ~threshold mem in
+            Addr.Set.union fs_s @@ Addr.Set.of_list new_fs_s
+          else fs_s) in
+  Symbols.write_addrset fs_set
+
+let find threshold length comp path print_sexp (input : string) : unit t =
   Image.create input >>= fun (img,_warns) ->
   let arch = Image.arch img in
   let data = Signatures.load ?path ~mode:"bytes" arch in
   Result.of_option data
     ~error:(Error.of_string "failed to read signatures from database")
   >>= fun data ->
-  let bw = Binable.of_string (module Byteweight) data in
-  Table.iteri (Image.sections img) ~f:(fun mem sec ->
-      if Section.is_executable sec then
-        let start = Memory.min_addr mem in
-        let rec loop n =
-          match Byteweight.next bw ~length ~threshold mem n with
-          | Some n -> printf "%a\n" Addr.ppo Addr.(start ++ n); loop (n+1)
-          | None -> () in
-        loop 0);
+  let bw = Binable.of_string (module BW) data in
+  (* if output sexp, get the addr set and output *)
+  if print_sexp then output_sexp img bw length threshold
+  else
+    Table.iteri (Image.sections img) ~f:(fun mem sec ->
+        if Section.is_executable sec then
+          let start = Memory.min_addr mem in
+          let rec loop n =
+            match BW.next bw ~length ~threshold mem n with
+            | Some n -> printf "%a\n" Addr.ppo Addr.(start ++ n); loop (n+1)
+            | None -> () in
+          loop 0);
   Ok ()
 
-let symbols print_name print_size input : unit t =
+let symbols print_name print_size print_sexp input : unit t =
```

Comment:
`print_sexp` breaks the logic. When you add `--print-sexp` it will ignore other options. This is not, what user expects.  Use expects, that if you do `--print-sexp --print-name --print-addr` all three options will be taken into account, and the sexp will contain only name and addr. So, indeed, you mixed orthogonal entities, columns of the table and output format. Instead of `--print-sexp` you should add `--output-format` option, that takes either `text` or `sexp`. And your sexp printer should respect `--print-*` options, i.e., print only whatever the user requested. Otherwise, if you think, that this behavior is to hard to implement, then create a new command, called `dump` that will dump data in specified format without any printing.